ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java

```diff
@@ -229,45 +265,20 @@ public AvroOutput(ValueProvider<String> outputPrefix, Duration windowDuration,
             "Path template must contain document namespace, type, and version");
       }
 
-      final PCollectionView<AvroSchemaStore> schemaSideInput = input.getPipeline()
-          .apply(Create.of(schemaStore)).apply(View.asSingleton());
-
-      final TupleTag<PubsubMessage> successTag = new TupleTag<PubsubMessage>() {
-      };
-      final TupleTag<PubsubMessage> errorTag = new TupleTag<PubsubMessage>() {
-      };
+      AvroEncoder encoder = new AvroEncoder();
 
       // A ParDo is opted over a PTransform extending MapElementsWithErrors.
       // While this leads to manual error handling with output-tags, this allows
       // for side-input of the singleton SchemaStore PCollection.
-      ParDo.MultiOutput<PubsubMessage, PubsubMessage> encodePayloadAsAvro = ParDo
-          .of(new DoFn<PubsubMessage, PubsubMessage>() {
-
-            @ProcessElement
-            public void processElement(ProcessContext ctx) throws SchemaNotFoundException {
-              PubsubMessage message = ctx.element();
-              Map<String, String> attributes = message.getAttributeMap();
-              try {
-                Schema schema = ctx.sideInput(schemaSideInput).getSchema(attributes);
-                GenericRecord record = formatter.formatRecord(message, schema);
-                byte[] avroPayload = binaryEncoder.encodeRecord(record, schema);
-                ctx.output(successTag, new PubsubMessage(avroPayload, attributes));
-              } catch (Exception e) {
-                ctx.output(errorTag, FailureMessage.of(this, message, e));
-              }
-            }
-          }).withSideInputs(schemaSideInput).withOutputTags(successTag, TupleTagList.of(errorTag));
+      ParDo.MultiOutput<PubsubMessage, PubsubMessage> encodePayloadAsAvro = ParDo.of(encoder)
+          .withOutputTags(successTag, TupleTagList.of(errorTag));
 
       FileIO.Write<List<String>, PubsubMessage> write = FileIO
           .<List<String>, PubsubMessage>writeDynamic() //
           .by(message -> pathTemplate.get().extractValuesFrom(message.getAttributeMap()))
           .withDestinationCoder(ListCoder.of(StringUtf8Coder.of())) //
           .withCompression(compression) //
-          .via(Contextful.fn((List<String> dest, Contextful.Fn.Context ctx) -> {
-            Map<String, String> attributes = pathTemplate.get().getPlaceholderAttributes(dest);
-            Schema schema = ctx.sideInput(schemaSideInput).getSchema(attributes);
-            return AvroIO.sinkViaGenericRecords(schema, binaryFormatter);
-          }, Requirements.requiresSideInputs(schemaSideInput))) //
+          .via(Contextful.fn(encoder::getSink)) //
```

Comment:
cc @acmiyaguchi re: changing this from a side input to a transient class variable